File: amp-web/src/drift.test.ts
/**
 * Wire-freeze drift guard — TypeScript side.
 *
 * Asserts the SAME golden fixtures the Go side round-trips
 * (amp.SDK/amp/webapi/testdata/*.json, guarded by amp/webapi/drift_test.go).
 * The Go structs are the wire spec; the shape tables below hand-list each
 * shape's keys AS GO DEFINES THEM — that hand-listing IS the drift guard,
 * since TS interfaces don't exist at runtime.  When a TS declaration in
 * types.ts disagrees with a table below, the TS side must move (Go is the
 * server); each known disagreement is flagged with a DRIFT note.
 *
 * Per fixture entry: required keys (Go fields without omitempty/omitzero)
 * must be present, and every present key must be known (required ∪ optional).
 * Nested wire objects are checked recursively.
 */

import { existsSync, readFileSync, readdirSync } from 'node:fs';
import { dirname, join, relative } from 'node:path';
import { fileURLToPath } from 'node:url';
import { describe, expect, it } from 'vitest';

import { CryptoKitID } from './crypto/types.js';
import { AmpWebClient } from './web-client.js';
import type { AccessLevel, InvitePolicyEntry, WithdrawReason } from './types.js';

// Fixture root, probed for both views this file ships in: the amp.SDK repo
// (amp/webapi/testdata) and the amp-web-SDK bundle (pack.sh stages the same
// fixtures at webapi/testdata).
const HERE = dirname(fileURLToPath(import.meta.url));
const TESTDATA_CANDIDATES = [
  join(HERE, '..', '..', 'amp', 'webapi', 'testdata'),
  join(HERE, '..', 'webapi', 'testdata'),
];
const testdataDir = TESTDATA_CANDIDATES.find(dir => existsSync(dir)) ?? '';
if (!testdataDir) {
  throw new Error(
    `wire-contract fixtures not found; looked in:\n  ${TESTDATA_CANDIDATES.join('\n  ')}`,
  );
}

function loadFixture(file: string): Record<string, unknown> {
  return JSON.parse(readFileSync(join(testdataDir, file), 'utf8'));
}

// ── Shape tables (the GO wire spec, hand-listed) ─────────────────────

interface ShapeSpec {
  required: readonly string[];
  optional?: readonly string[];
  /** wire-object sub-shapes by key; arrays are checked element-wise */
  nested?: Record<string, string>;
}

const SHAPES: Record<string, ShapeSpec> = {
  // types.ts LoginCredentials DELIBERATELY omits the optional PlanetTag the
  // Go shape accepts: the TS client performs no login-planet bind — the
  // constructor planetTag rides each call instead, and the server stays sole
  // authority on planet resolution.
  LoginRequest: {
    required: ['Scheme'],
    optional: ['Address', 'Signature', 'Nonce', 'Email', 'Password',
      'MemberToken', 'ChallengeResponse', 'DID', 'PlanetTag'],
  },
  LoginResponse: {
    required: ['SessionToken', 'ExpiresAt', 'Member'],
    nested: { Member: 'AmpMember' },
  },
  // DisplayName is optional on the wire (Go omitempty; types.ts matches).
  // Go is the spec — DisplayName is optional here; types.ts should follow.
  AmpMember: {
    required: ['ID', 'PlanetID'],
    optional: ['DisplayName', 'Email', 'Kind', 'Address'],
  },
  SessionResponse: {
    required: ['Member', 'ExpiresAt'],
    nested: { Member: 'AmpMember' },
  },
  EmailCredential: {
    required: [],
    optional: ['Email', 'Password', 'Token', 'NewPassword', 'PlanetTag'],
  },
  EmailIssueResponse: { required: ['MemberID', 'Email'] },
  ErrorResponse: { required: ['Code', 'Message'] },

  WithdrawNote: {
    required: ['Reason'],
    optional: ['Rationale', 'Subject', 'Delegation', 'WithdrawnAt', 'WithdrawnBy'],
  },
  TxOp: {
    required: ['Kind', 'Channel', 'Attr'],
    optional: ['ItemID', 'Value', 'Withdraw'],
    nested: { Withdraw: 'WithdrawNote' },
  },
  TxRequest: {
    required: ['Ops'],
    optional: ['PlanetTag', 'InvokeURL'],
    nested: { Ops: 'TxOp' },
  },
  TxOpResult: { required: ['ItemID', 'EditID'], optional: ['Error'] },
  TxResponse: {
    required: ['TxID', 'Results'],
    nested: { Results: 'TxOpResult' },
  },
  Item: {
    required: ['_ItemID', '_EditID', '_FromID', '_UpdatedAt', 'Value'],
    optional: ['_Withdrawn'],
    nested: { _Withdrawn: 'WithdrawNote' },
  },
  ListResponse: {
    required: ['Items', 'HasMore'],
    optional: ['Next'],
    nested: { Items: 'Item' },
  },

  SubscribeFrame: {
    required: ['Type'],
    optional: ['Channel', 'Attr', 'ItemID', 'EditID', 'FromID', 'Value',
      'UpdatedAt', 'Withdraw', 'Error'],
    nested: { Withdraw: 'WithdrawNote' },
  },

  // NOTE: no TS interface models EditEntry / EditChainResponse yet (the SDK
  // has no edit-chain API); these assert the Go shape so the fixture stays
  // honest for the client that adds it.
  EditEntry: {
    required: ['EditID', 'CommitTx', 'Author', 'CommittedAt', 'Op'],
    optional: ['Withdraw', 'Body'],
    nested: { Withdraw: 'WithdrawNote' },
  },
  EditChainResponse: {
    required: ['Edits'],
    optional: ['Original'],
    nested: { Original: 'Item', Edits: 'EditEntry' },
  },

  InviteIssueRequest: {
    required: ['Planet', 'Passphrase'],
    optional: ['MaxRedemptions', 'Access', 'ExpiresAt', 'VaultAddrs'],
  },
  InviteIssueResponse: { required: ['PlanetID', 'InviteID', 'InviteText'] },
  InviteAcceptRequest: { required: ['InviteText', 'Passphrase'] },
  InviteAcceptResponse: { required: ['PlanetID', 'MemberID'] },
  InviteRevokeRequest: {
    required: ['Planet'],
    optional: ['InviteID', 'InviteText', 'Rotate'],
  },
  InviteListResponse: {
    required: ['Policies'],
    nested: { Policies: 'InvitePolicyEntry' },
  },
  InvitePolicyEntry: {
    required: ['InviteID', 'MaxRedemptions', 'Status'],
    optional: ['GrantedAccess', 'ExpiresAt', 'Redemptions'],
    nested: { Redemptions: 'InviteRedemptionEntry' },
  },
  InviteRedemptionEntry: { required: ['Member', 'RedeemedAt', 'Rank', 'InRank'] },

  // Address is BASE64 — opaque transport bytes (Go []byte), NOT a base32 UID.
  // No TS interface models the resolve/federation family yet.
  VaultEndpoint: { required: ['Transport', 'Address'] },
  ResolveRequest: { required: ['FQDN'] },
  ResolveResponse: {
    required: ['FQDN', 'PlanetID', 'AnsweredBy', 'TrustState',
      'PinPrecedence', 'Ambiguous', 'Hops'],
    optional: ['VaultAddrs'],
    nested: { VaultAddrs: 'VaultEndpoint' },
  },
  SearchRequest: { required: ['Query'], optional: ['Limit'] },
  SearchResponse: {
    required: ['Matches'],
    nested: { Matches: 'SearchMatch' },
  },
  SearchMatch: {
    required: ['PlanetID', 'FQDN', 'AnsweredBy', 'Score', 'AppName', 'AppDesc'],
    optional: ['Platforms'],
  },
  FederationPeersResponse: {
    required: ['Peers'],
    nested: { Peers: 'FederationPeerEntry' },
  },
  FederationPeerEntry: {
    required: ['FederationID'],
    optional: ['VaultAddrs', 'Label'],
    nested: { VaultAddrs: 'VaultEndpoint' },
  },

  MediaResolveRequest: {
    required: ['Blob'],
    optional: ['PlanetTag'],
    nested: { Blob: 'Tag' },
  },
  // amp.Tag JSON (Go tagJSON): every field omitempty.
  // The wire key is ContentTypeRaw (types.ts BlobRef matches).
  Tag: {
    required: [],
    optional: ['UID', 'I', 'J', 'K', 'Units', 'ContentTypeRaw', 'URI', 'Text'],
  },

  // Operator tier — Go-only by ruling (see OPERATOR_GO_ONLY below): these
  // assert the Go wire shape so the fixtures stay honest, and deliberately
  // have NO types.ts interface and NO client method.
  PlanetCreateRequest: {
    required: [],
    optional: ['Tag', 'Brand'],
    nested: { Tag: 'Tag', Brand: 'Brand' },
  },
  PlanetCreateResponse: {
    required: [],
    optional: ['PlanetID', 'Tag', 'Public'],
  },
  BrandSetRequest: {
    required: [],
    optional: ['Planet', 'Brand'],
    nested: { Planet: 'Tag', Brand: 'Brand' },
  },
  BrandSetResponse: {
    required: [],
    optional: ['PlanetID'],
  },
  ForumsReserveRequest: {
    required: [],
    optional: ['Address', 'MemberID'],
  },
  ForumsReserveResponse: { required: ['MemberID'] },
  // amp.Brand / amp.BrandIdentity (proto-generated; every field omitempty).
  // Targets/Links/BundledCrates elements are unexercised by fixtures today.
  Brand: {
    required: [],
    optional: ['Identity', 'OrgHomeURL', 'AppHomeURL', 'Targets',
      'CrateSnapshotURL', 'Links', 'BundledCrates', 'TemplateSet'],
    nested: { Identity: 'BrandIdentity', TemplateSet: 'Tag' },
  },
  BrandIdentity: {
    required: [],
    optional: ['AppName', 'OrgName', 'AppDomain', 'AppDesc', 'URLSchemes', 'NamedBy'],
    nested: { NamedBy: 'Tag' },
  },
};

// Keys carrying app-defined payloads (json.RawMessage Go-side) — any inner
// keys are legal, so they are exempt from nested checks by construction
// (absent from `nested` above): TxOp.Value, Item.Value, SubscribeFrame.Value,
// EditEntry.Body.

const FIXTURE_FILES = [
  'login.json',
  'tx-item.json',
  'subscribe.json',
  'edits.json',
  'invite.json',
  'vault.json',
  'media.json',
  'operator.json',
] as const;

const shapesChecked = new Set<string>();

function checkShape(value: unknown, shapeName: string, path: string): void {
  const spec = SHAPES[shapeName];
  expect(spec, `${path}: unknown shape ${shapeName} — add it to SHAPES`).toBeDefined();
  shapesChecked.add(shapeName);

  expect(value, `${path}: expected a wire object`).toBeTypeOf('object');
  expect(value, `${path}: expected a wire object, got null`).not.toBeNull();
  const obj = value as Record<string, unknown>;

  const known = new Set([...spec.required, ...(spec.optional ?? [])]);
  for (const key of spec.required) {
    expect(
      Object.prototype.hasOwnProperty.call(obj, key),
      `${path}: required key ${shapeName}.${key} missing from fixture`,
    ).toBe(true);
  }
  for (const key of Object.keys(obj)) {
    expect(
      known.has(key),
      `${path}: fixture key ${shapeName}.${key} is not in the ${shapeName} shape table — wire drift or stale table`,
    ).toBe(true);
  }

  for (const [key, subShape] of Object.entries(spec.nested ?? {})) {
    const sub = obj[key];
    if (sub === undefined || sub === null) continue;
    if (Array.isArray(sub)) {
      sub.forEach((el, i) => checkShape(el, subShape, `${path}.${key}[${i}]`));
    } else {
      checkShape(sub, subShape, `${path}.${key}`);
    }
  }
}

// ── Structural fixture checks ────────────────────────────────────────

describe('wire fixtures match the Go shapes', () => {
  for (const file of FIXTURE_FILES) {
    it(file, () => {
      const entries = loadFixture(file);
      expect(Object.keys(entries).length).toBeGreaterThan(0);
      for (const [name, value] of Object.entries(entries)) {
        const shape = name.split('.')[0];
        checkShape(value, shape, `${file}#${name}`);
      }
    });
  }

  it('every shape table is exercised by a fixture', () => {
    for (const shape of Object.keys(SHAPES)) {
      expect(shapesChecked.has(shape), `shape ${shape} never checked — fixture entry deleted?`).toBe(true);
    }
  });
});

// ── OPERATOR_GO_ONLY — the operator tier has NO client binding ──────
//
// testdata/operator-go-only.json lists the /api/v1/admin/* verbs ruled
// Go/CLI-only: the operator Bearer is higher-privilege than a member session
// and must never normalize into XSS-exposed browser JS.  The Go side pins the
// same manifest's shapes into its drift guard; this side asserts the INVERSE —
// no AmpWebClient method for the verb, no admin endpoint string anywhere in
// client source.  Adding a TS binding therefore fails here until the manifest
// itself is edited (a reviewed act).

interface OperatorVerb {
  Verb: string;
  Endpoint: string;
  Request: string;
  Response: string;
}

const OPERATOR_GO_ONLY = (loadFixture('operator-go-only.json') as unknown as {
  Verbs: OperatorVerb[];
}).Verbs;

/** Every non-test source file under src/ (the shipped client surface). */
function clientSourceFiles(): string[] {
  const srcDir = dirname(fileURLToPath(import.meta.url));
  return readdirSync(srcDir, { recursive: true, encoding: 'utf8' })
    .filter(f => /\.(ts|tsx)$/.test(f) && !/\.test\.ts$/.test(f) && !f.endsWith('.d.ts'))
    .map(f => join(srcDir, f));
}

describe('OPERATOR_GO_ONLY manifest — no client binding', () => {
  it('manifest lists the operator tier', () => {
    expect(OPERATOR_GO_ONLY.length).toBeGreaterThan(0);
  });

  it('every operator shape is drift-guarded (present in SHAPES)', () => {
    for (const verb of OPERATOR_GO_ONLY) {
      for (const shape of [verb.Request, verb.Response]) {
        expect(SHAPES[shape], `operator shape ${shape} missing from SHAPES`).toBeDefined();
      }
    }
  });

  it('AmpWebClient exposes no method for an operator verb', () => {
    const methods = new Set(Object.getOwnPropertyNames(AmpWebClient.prototype));
    for (const verb of OPERATOR_GO_ONLY) {
      expect(
        methods.has(verb.Verb),
        `AmpWebClient.${verb.Verb}() exists — operator verbs are Go-only; edit operator-go-only.json only via review`,
      ).toBe(false);
    }
  });

  it('no client CODE references an operator admin endpoint', () => {
    const srcDir = dirname(fileURLToPath(import.meta.url));
    for (const file of clientSourceFiles()) {
      // Strip comments — an endpoint may be DOCUMENTED (e.g. the shared
      // EmailCredential shape lists its three endpoints), never called.
      const code = readFileSync(file, 'utf8')
        .replace(/\/\*[\s\S]*?\*\//g, '')
        .replace(/^\s*\/\/.*$/gm, '');
      for (const verb of OPERATOR_GO_ONLY) {
        expect(
          code.includes(verb.Endpoint),
          `${relative(srcDir, file)} references ${verb.Endpoint} — operator verbs are Go-only`,
        ).toBe(false);
      }
    }
  });
});

// ── Enum-name goldens ────────────────────────────────────────────────

// `satisfies` ties each hand-list to the TS union at typecheck time; the
// runtime toEqual ties it to the shared fixture the Go side verifies.
const ACCESS_LEVELS = [
  'ReadOnly', 'ReadWrite', 'Moderator', 'Admin',
] as const satisfies readonly AccessLevel[];

const WITHDRAW_REASONS = [
  'Consent', 'Inaccuracy', 'Outdated', 'Coerced',
  'Forgotten', 'Departed', 'InviteRecall', 'Retracted',
] as const satisfies readonly WithdrawReason[];

const INVITE_STATUSES = [
  'InviteActive', 'InviteRevoked',
] as const satisfies readonly InvitePolicyEntry['Status'][];

describe('enum-name goldens', () => {
  const golden = loadFixture('access.json') as {
    AccessLevels: string[];
    WithdrawReasons: string[];
    InviteStatuses: string[];
    TrustStates: string[];
  };

  it('AccessLevel union matches the golden', () => {
    expect(golden.AccessLevels).toEqual([...ACCESS_LEVELS]);
  });

  it('WithdrawReason union matches the golden', () => {
    expect(golden.WithdrawReasons).toEqual([...WITHDRAW_REASONS]);
  });

  it('InvitePolicyEntry.Status union matches the golden', () => {
    expect(golden.InviteStatuses).toEqual([...INVITE_STATUSES]);
  });

  // TrustStates golden is Go-verified only: no TS type models the resolve
  // family yet (see ResolveResponse note above).
});

// ── CryptoKitID golden ───────────────────────────────────────────────

describe('CryptoKitID table matches the golden', () => {
  const golden = loadFixture('cryptokits.json') as {
    Kits: { Name: string; TsKitID: number; GoUID: string }[];
  };

  it('every golden kit maps to the TS small-int', () => {
    const table = CryptoKitID as Record<string, number>;
    for (const kit of golden.Kits) {
      expect(table[kit.Name], `kit ${kit.Name} missing from CryptoKitID`).toBe(kit.TsKitID);
    }
  });

  it('every TS kit appears in the golden', () => {
    const goldenNames = new Set(golden.Kits.map(kit => kit.Name));
    for (const name of Object.keys(CryptoKitID)) {
      expect(goldenNames.has(name), `CryptoKitID.${name} missing from the golden`).toBe(true);
    }
  });

  it('Go kit UIDs are well-formed base32 (authoritative values asserted Go-side)', () => {
    for (const kit of golden.Kits) {
      if (kit.Name === 'Unspecified') {
        expect(kit.GoUID).toBe('');
        continue;
      }
      expect(kit.GoUID).toMatch(/^[0-9a-z]{6}(?:-[0-9a-z]{5}){4}$/);
    }
  });
});

File: amp/webapi/drift_test.go
package webapi_test

// Wire-freeze drift guard.  testdata/*.json are the shared golden fixtures for
// the /api/v1/* JSON contract — the SAME files are asserted by the TypeScript
// SDK (amp-web/src/drift.test.ts), so a silent field rename, json-tag edit, or
// omitempty/omitzero regression on either side fails that side's CI.
//
// Per fixture entry: decode into the Go wire struct with DisallowUnknownFields
// (catches removed/renamed Go fields), re-marshal, and require the two JSON
// trees to be deeply equal as maps (catches dropped populated keys AND
// newly-emitted zero-value keys, order-insensitive).  Fixture entry names are
// "{Shape}" or "{Shape}.{variant}" — the prefix picks the Go type.

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"reflect"
	"sort"
	"strings"
	"testing"

	"github.com/art-media-platform/amp.SDK/amp"
	"github.com/art-media-platform/amp.SDK/amp/webapi"
	"github.com/art-media-platform/amp.SDK/stdlib/safe"
)

// nonShapeFiles hold enum-name goldens and the operator manifest, not struct
// shapes; each has its own test below.
var nonShapeFiles = map[string]bool{
	"access.json":           true,
	"cryptokits.json":       true,
	"operator-go-only.json": true,
}

// shapeCtors maps a fixture entry's shape prefix to its Go wire struct.
var shapeCtors = map[string]func() any{
	"LoginRequest":            func() any { return &webapi.LoginRequest{} },
	"ChallengeResponse":       func() any { return &webapi.ChallengeResponse{} },
	"LoginResponse":           func() any { return &webapi.LoginResponse{} },
	"AmpMember":               func() any { return &webapi.AmpMember{} },
	"SessionResponse":         func() any { return &webapi.SessionResponse{} },
	"EmailCredential":         func() any { return &webapi.EmailCredential{} },
	"EmailIssueResponse":      func() any { return &webapi.EmailIssueResponse{} },
	"ErrorResponse":           func() any { return &webapi.ErrorResponse{} },
	"WithdrawNote":            func() any { return &webapi.WithdrawNote{} },
	"TxOp":                    func() any { return &webapi.TxOp{} },
	"TxRequest":               func() any { return &webapi.TxRequest{} },
	"TxOpResult":              func() any { return &webapi.TxOpResult{} },
	"TxResponse":              func() any { return &webapi.TxResponse{} },
	"Item":                    func() any { return &webapi.Item{} },
	"ListResponse":            func() any { return &webapi.ListResponse{} },
	"SubscribeFrame":          func() any { return &webapi.SubscribeFrame{} },
	"EditEntry":               func() any { return &webapi.EditEntry{} },
	"EditChainResponse":       func() any { return &webapi.EditChainResponse{} },
	"InviteIssueRequest":      func() any { return &webapi.InviteIssueRequest{} },
	"InviteIssueResponse":     func() any { return &webapi.InviteIssueResponse{} },
	"InviteAcceptRequest":     func() any { return &webapi.InviteAcceptRequest{} },
	"InviteAcceptResponse":    func() any { return &webapi.InviteAcceptResponse{} },
	"InviteRevokeRequest":     func() any { return &webapi.InviteRevokeRequest{} },
	"InviteListResponse":      func() any { return &webapi.InviteListResponse{} },
	"GovernanceGrantRequest":  func() any { return &webapi.GovernanceGrantRequest{} },
	"GovernanceGrantResponse": func() any { return &webapi.GovernanceGrantResponse{} },
	"VaultEndpoint":           func() any { return &webapi.VaultEndpoint{} },
	"ResolveRequest":          func() any { return &webapi.ResolveRequest{} },
	"ResolveResponse":         func() any { return &webapi.ResolveResponse{} },
	"SearchRequest":           func() any { return &webapi.SearchRequest{} },
	"SearchResponse":          func() any { return &webapi.SearchResponse{} },
	"FederationPeersResponse": func() any { return &webapi.FederationPeersResponse{} },
	"MediaResolveRequest":     func() any { return &webapi.MediaResolveRequest{} },
	"Tag":                     func() any { return &amp.Tag{} },

	// Operator tier (Go-only by design — see testdata/operator-go-only.json;
	// TestOperatorGoOnlyManifest pins these registrations to that manifest).
	"PlanetCreateRequest":   func() any { return &webapi.PlanetCreateRequest{} },
	"PlanetCreateResponse":  func() any { return &webapi.PlanetCreateResponse{} },
	"BrandSetRequest":       func() any { return &webapi.BrandSetRequest{} },
	"BrandSetResponse":      func() any { return &webapi.BrandSetResponse{} },
	"ForumsReserveRequest":  func() any { return &webapi.ForumsReserveRequest{} },
	"ForumsReserveResponse": func() any { return &webapi.ForumsReserveResponse{} },
}

func TestWireFixturesRoundTrip(t *testing.T) {
	files, err := filepath.Glob(filepath.Join("testdata", "*.json"))
	if err != nil {
		t.Fatal(err)
	}
	if len(files) == 0 {
		t.Fatal("no fixtures under testdata/")
	}

	shapesSeen := map[string]bool{}
	for _, file := range files {
		if nonShapeFiles[filepath.Base(file)] {
			continue
		}
		raw, err := os.ReadFile(file)
		if err != nil {
			t.Fatal(err)
		}
		entries := map[string]json.RawMessage{}
		if err := json.Unmarshal(raw, &entries); err != nil {
			t.Fatalf("%s: %v", file, err)
		}

		names := make([]string, 0, len(entries))
		for name := range entries {
			names = append(names, name)
		}
		sort.Strings(names)

		for _, name := range names {
			shape, _, _ := strings.Cut(name, ".")
			ctor := shapeCtors[shape]
			if ctor == nil {
				t.Errorf("%s: fixture %q has no Go shape registered — add it to shapeCtors", file, name)
				continue
			}
			shapesSeen[shape] = true
			roundTrip(t, fmt.Sprintf("%s: %s", filepath.Base(file), name), entries[name], ctor())
		}
	}

	// Reverse coverage: a registered shape with no fixture means a fixture
	// entry was deleted (or renamed) without updating this registry.
	for shape := range shapeCtors {
		if !shapesSeen[shape] {
			t.Errorf("shape %s registered but no fixture entry exercises it", shape)
		}
	}
}

// roundTrip decodes the fixture into target rejecting unknown fields, then
// re-marshals and requires deep JSON-tree equality with the fixture.
func roundTrip(t *testing.T, label string, fixture json.RawMessage, target any) {
	dec := json.NewDecoder(bytes.NewReader(fixture))
	dec.DisallowUnknownFields()
	if err := dec.Decode(target); err != nil {
		t.Errorf("%s: decode into %T: %v", label, target, err)
		return
	}

	remarshaled, err := json.Marshal(target)
	if err != nil {
		t.Errorf("%s: re-marshal %T: %v", label, target, err)
		return
	}

	var want, got map[string]any
	if err := json.Unmarshal(fixture, &want); err != nil {
		t.Errorf("%s: fixture is not a JSON object: %v", label, err)
		return
	}
	if err := json.Unmarshal(remarshaled, &got); err != nil {
		t.Errorf("%s: re-marshaled %T is not a JSON object: %v", label, target, err)
		return
	}
	if reflect.DeepEqual(want, got) {
		return
	}

	// Actionable diff: per-key missing / extra / changed.
	for key, wantVal := range want {
		gotVal, ok := got[key]
		switch {
		case !ok:
			t.Errorf("%s: key %q dropped by round trip (omitempty/omitzero or json-tag drift?)", label, key)
		case !reflect.DeepEqual(wantVal, gotVal):
			t.Errorf("%s: key %q changed by round trip:\n  fixture:    %v\n  remarshal:  %v", label, key, wantVal, gotVal)
		}
	}
	for key := range got {
		if _, ok := want[key]; !ok {
			t.Errorf("%s: round trip emitted key %q absent from the fixture (lost omitempty/omitzero, or fixture is missing a required field)", label, key)
		}
	}
}

// TestOperatorGoOnlyManifest pins the operator-tier manifest: every listed
// verb's request/response shape must be registered in shapeCtors (and so
// fixture-guarded by TestWireFixturesRoundTrip's reverse coverage).  The TS
// side asserts the inverse against the same file — that the browser client
// has no binding for these verbs (NodeAdminModule=true verbs bind only in the
// Node-only admin module) — so an operator verb cannot silently gain a
// browser method or silently fall out of the drift guard.
func TestOperatorGoOnlyManifest(t *testing.T) {
	manifest := struct {
		Comment string
		Verbs   []struct {
			Verb            string
			Endpoint        string
			Request         string
			Response        string
			NodeAdminModule bool
		}
	}{}
	loadJSON(t, filepath.Join("testdata", "operator-go-only.json"), &manifest)

	if len(manifest.Verbs) == 0 {
		t.Fatal("operator manifest lists no verbs")
	}
	for _, verb := range manifest.Verbs {
		if verb.Verb == "" || !strings.HasPrefix(verb.Endpoint, "/api/v1/admin/") {
			t.Errorf("manifest verb %q: endpoint %q is not an /api/v1/admin/ surface", verb.Verb, verb.Endpoint)
		}
		for _, shape := range []string{verb.Request, verb.Response} {
			if shapeCtors[shape] == nil {
				t.Errorf("manifest verb %q: shape %s not registered in shapeCtors — operator shape outside the drift guard", verb.Verb, shape)
			}
		}
	}
}

// TestAccessGolden pins the enum NAMES that ride the JSON wire.  The TS side
// asserts the same file against its string-literal unions.
func TestAccessGolden(t *testing.T) {
	golden := struct {
		AccessLevels    []string
		WithdrawReasons []string
		InviteStatuses  []string
		TrustStates     []string
	}{}
	loadJSON(t, filepath.Join("testdata", "access.json"), &golden)

	// Every golden mirrors its full wire vocabulary — set-equal, so adding an
	// enum value Go-side forces the golden (and the TS union) to move.
	// AccessLevels is the full amp.Access set: the grant endpoints (invite
	// issue, governance grant) parse any Access enum name, and NotAllowed is
	// meaningful as an explicit-deny governance grant.
	assertNameSetEqual(t, "Access", golden.AccessLevels, amp.Access_name)
	assertNameSetEqual(t, "WithdrawReason", golden.WithdrawReasons, amp.WithdrawReason_name, "UnspecifiedReason")
	assertNameSetEqual(t, "InviteStatus", golden.InviteStatuses, amp.InviteStatus_name)
	assertNameSetEqual(t, "TrustState", golden.TrustStates, amp.TrustState_name)
}

// assertNameSetEqual requires golden to equal the proto enum's name set,
// minus any excluded names (invalid-on-the-wire defaults).
func assertNameSetEqual(t *testing.T, enum string, golden []string, names map[int32]string, exclude ...string) {
	excluded := map[string]bool{}
	for _, name := range exclude {
		excluded[name] = true
	}
	want := map[string]bool{}
	for _, name := range names {
		if !excluded[name] {
			want[name] = true
		}
	}
	for _, name := range golden {
		if !want[name] {
			t.Errorf("%s golden %q is not a wire-valid Go enum name", enum, name)
		}
		delete(want, name)
	}
	for name := range want {
		t.Errorf("%s enum name %q missing from the golden (added Go-side without updating the contract?)", enum, name)
	}
}

// TestCryptoKitsGolden pins the TS CryptoKitID small-int table to the Go
// safe.Crypto tag.UID registry.  The TS side asserts the same file against
// its CryptoKitID const table; the GoUID column is authoritative here.
func TestCryptoKitsGolden(t *testing.T) {
	golden := struct {
		Kits []struct {
			Name    string
			TsKitID int
			GoUID   string
		}
	}{}
	loadJSON(t, filepath.Join("testdata", "cryptokits.json"), &golden)

	goKits := map[string]string{
		"Poly25519": safe.Crypto.Poly25519.ID.String(),
		"P256":      safe.Crypto.P256.ID.String(),
		"Secp256k1": safe.Crypto.Secp256k1.ID.String(),
	}

	for _, kit := range golden.Kits {
		if kit.Name == "Unspecified" {
			if kit.GoUID != "" {
				t.Errorf("Unspecified kit must carry no Go UID, got %q", kit.GoUID)
			}
			continue
		}
		want, ok := goKits[kit.Name]
		if !ok {
			t.Errorf("kit golden %q has no safe.Crypto counterpart", kit.Name)
			continue
		}
		if kit.GoUID != want {
			t.Errorf("kit %s: golden GoUID %q != safe.Crypto UID %q", kit.Name, kit.GoUID, want)
		}
		delete(goKits, kit.Name)
	}
	for name := range goKits {
		t.Errorf("safe.Crypto kit %s missing from the golden", name)
	}
}

func loadJSON(t *testing.T, path string, target any) {
	t.Helper()
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.DisallowUnknownFields()
	if err := dec.Decode(target); err != nil {
		t.Fatalf("%s: %v", path, err)
	}
}
